File: 지현배/2108/0826/21738.py
import sys
from collections import deque
input = sys.stdin.readline
N, S, P = map(int, input().split())
connections = [list(map(int, input().split())) for _ in range(N - 1)]
link = [[] for _ in range(N + 1)]
# 해당 인덱스에 연결된 얼음 인덱스들을 넣는다.
for s, e in connections:
    link[s].append(e)
    link[e].append(s)
answer = 0
life = 2
queue = deque([[P, 0, -1]])
while queue:
    curr, cnt, prev = queue.popleft()
    # 종료 조건: 도착한 얼음이 지지 얼음이면 그때까지의 카운트를 답에 더하고
    # 라이프를 하나 깍는다. 그 라이프가 0이 되면 종료한다.
    if curr <= S:
        answer += cnt
        life -= 1
        if life <= 0: break
    else:
        # 해당 얼음에서 갈 수 있는 얼음 중 이전 것을 제외하고 다음 얼음을 큐에 넣는다.
        next = link[curr]
        for e in next:
            if e != prev:
                queue.append([e, cnt + 1, curr])
# 총 얼음에서 저장된 값과 펭귄이 올라가 있는 값 1을 뺀 값을 출력한다.
print(N - answer - 1)

File: 지현배/2108/0826/21738.js
// const fs = require('fs');
// let [NSP, ...connections] = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
let NSP = '21 6 12';
let connections = [
  '1 9',
  '1 10',
  '10 12',
  '2 13',
  '13 11',
  '11 12',
  '3 8',
  '8 7',
  '8 12',
  '5 19',
  '5 14',
  '14 12',
  '6 20',
  '6 21',
  '20 15',
  '15 12',
  '4 18',
  '4 17',
  '17 16',
  '16 12'
];
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}
class Queue {
  constructor() {
    this.head = null;
    this.tail = null;
    this.size = 0;
  }
  enQueue(value) {
    let node = new Node(value);
    if (this.head == null) {
      this.head = node;
      this.head.next = this.tail;
    }
    else this.tail.next = node;
    this.tail = node;
    this.size++;
  }
  deQueue() {
    let ret = this.head;
    this.head = this.head.next;
    this.size--;
    return ret;
  }
  getSize() {
    return this.size;
  }
}
let [N, S, P] = NSP.split(' ').map(Number);
let link = Array(N + 1).fill().map(e => []);
connections.map(el => {
  let [s, e] = el.split(' ').map(Number);
  link[s].push(e);
  link[e].push(s);
});
let answer = 0;
let life = 2;
let queue = new Queue();
queue.enQueue([P, 0, -1]); // 현위치, 카운트, 전위치
while (queue.getSize() > 0) {
  let [curr, cnt, prev] = queue.deQueue().value;
  if (curr <= S) {
    answer += cnt;
    life--;
    if (life <= 0) break;
  }
  else {
    let next = link[curr];
    next.map(e => {
      if (e != prev) {
        queue.enQueue([e, cnt + 1, curr]);
      }
    });
  }
}
console.log(N - answer - 1)